src/RequestHandler/inprocess/inprocessapplication.cpp

```diff
@@ -30,99 +30,121 @@ IN_PROCESS_APPLICATION::~IN_PROCESS_APPLICATION()
     Recycle();
 }
 
-__override
 VOID
-IN_PROCESS_APPLICATION::ShutDown()
+IN_PROCESS_APPLICATION::Recycle()
 {
-    //todo
+
+    HANDLE   handle = NULL;
+    WIN32_FIND_DATA fileData;
+
+    if (!m_pHttpServer->IsCommandLineLaunch() &&
+        (m_pHttpServer->GetAdminManager() != NULL))
+    {
+        // IIS scenario.
+        // notify IIS first so that new request will be routed to new worker process
+        m_pHttpServer->RecycleProcess(L"AspNetCore Recycle Process on Demand");
+    }
+
+    s_Application = NULL;
+
+    if (m_pStdFile != NULL)
+    {
+        fflush(stdout);
+        fflush(stderr);
+        fclose(m_pStdFile);
+    }
+
+    if (m_hLogFileHandle != INVALID_HANDLE_VALUE)
+    {
+        m_Timer.CancelTimer();
+        CloseHandle(m_hLogFileHandle);
+        m_hLogFileHandle = INVALID_HANDLE_VALUE;
+    }
+
+    // delete empty log file, if logging is not enabled
+    handle = FindFirstFile(m_struLogFilePath.QueryStr(), &fileData);
+    if (handle != INVALID_HANDLE_VALUE &&
+        fileData.nFileSizeHigh == 0 &&
+        fileData.nFileSizeLow == 0) // skip check of nFileSizeHigh
+    {
+        FindClose(handle);
+        // no need to check whether the deletion succeeds
+        // as nothing can be done
+        DeleteFile(m_struLogFilePath.QueryStr());
+    }
+
+    if (m_pHttpServer && m_pHttpServer->IsCommandLineLaunch())
+    {
+        // IISExpress scenario
+        // Can only call exit to terminate current process
+        exit(0);
+    }
 }
 
-// This is the same function as before, TODO configrm if we need to change anything for configuration.
+__override
 VOID
-IN_PROCESS_APPLICATION::Recycle(
-    VOID
-)
+IN_PROCESS_APPLICATION::ShutDown()
 {
-    if (m_fInitialized)
+    DWORD    dwTimeout;
+    DWORD    dwThreadStatus = 0;
+    BOOL     fLocked = FALSE;
+
+    if (IsDebuggerPresent())
+    {
+        dwTimeout = INFINITE;
+    }
+    else
     {
-        DWORD    dwThreadStatus = 0;
-        DWORD    dwTimeout = m_pConfig->QueryShutdownTimeLimitInMS();
-        HANDLE   handle = NULL;
-        WIN32_FIND_DATA fileData;
+        dwTimeout = m_pConfig->QueryShutdownTimeLimitInMS();
+    }
 
-        AcquireSRWLockExclusive(&m_srwLock);
+    if (m_fShutdownCalledFromNative || !m_fManagedAppLoaded)
+    {
+        goto Finished;
+    }
+    AcquireSRWLockExclusive(&m_srwLock);
+    fLocked = TRUE;
+    if (m_fShutdownCalledFromNative || !m_fManagedAppLoaded)
+    {
+        goto Finished;
+    }
 
-        if (!m_pHttpServer->IsCommandLineLaunch() &&
-            !m_fRecycleProcessCalled &&
-            (m_pHttpServer->GetAdminManager() != NULL))
-        {
-            // IIS scenario.
-            // notify IIS first so that new request will be routed to new worker process
-            m_pHttpServer->RecycleProcess(L"AspNetCore Recycle Process on Demand");
-        }
+    m_fShutdownCalledFromNative = TRUE;
```

Comment:
why you have m_fShutdownCalledFromNative = TRUE while still check !m_fshutdwonCalledFromManaged?
hot could a shutdown is initiated from managed?

Reply:
Shutdown from managed can happened either when Main returns or when the server calls IApplicationLifetime.Shutdown(), which will trigger the managed server shutdown.